kernel/src/checkpoint/tests.rs

```diff
@@ -505,3 +506,81 @@ fn test_v2_checkpoint_supported_table() -> DeltaResult<()> {
 
     Ok(())
 }
+
+/// unpack the test data tarball from `path` into a temp dir, and return the dir it was
+/// unpacked into
+#[allow(unused)]
+pub(crate) fn load_test_data(
+    test_parent_dir: &str,
+    test_name: &str,
+) -> Result<tempfile::TempDir, Box<dyn std::error::Error>> {
+    let path = format!("{test_parent_dir}/{test_name}.tar.zst");
+    let tar = zstd::Decoder::new(std::fs::File::open(path)?)?;
+    let mut archive = tar::Archive::new(tar);
+    let temp_dir = tempfile::tempdir()?;
+    archive.unpack(temp_dir.path())?;
+    Ok(temp_dir)
+}
+
+/// Tests the `checkpoint()` API with:
+/// - A table that does supports v2Checkpoint
+/// - No version specified (latest version is used)
+#[test]
+fn two_phase_checkpoint_replay() {
+    let test_name = "v2-checkpoints-parquet-with-sidecars";
+    let tmp_path = load_test_data("tests/data", test_name).unwrap();
+    let table_path = tmp_path.path().join(test_name);
+    let log_path = table_path.join("_delta_log");
+
+    // The table has a checkpoint for every commit (v6 is latest), which makes it hard to fully test
+    // two-phase log replay. Manually delete checkpoints 5 and 6, so we replay deltas for v6 and v5
+    // on top of the checkpoint for v4. Also delete the (too new) _last_checkpoint file.
+    std::fs::remove_file(
+        log_path
+            .join("00000000000000000005.checkpoint.cda8d476-9f57-4c68-866b-0bcdd4bfe71b.parquet"),
+    )
+    .unwrap();
+    std::fs::remove_file(
+        log_path
+            .join("00000000000000000006.checkpoint.f15b9025-707a-4c73-aac0-31dfcbd29aa6.parquet"),
+    )
+    .unwrap();
+    std::fs::remove_file(log_path.join("_last_checkpoint")).unwrap();
+    let engine = crate::engine::sync::SyncEngine::new();
+    let snapshot = Snapshot::try_from_uri(
+        table_path.to_str().expect("table path to string"),
+        &engine,
+        None,
+    )
+    .unwrap();
+
+    // Traditional log replay
+    let checkpoint_writer = Arc::new(snapshot).checkpoint_writer().unwrap();
+    let mut checkpoint_data = checkpoint_writer.checkpoint_data(&engine).unwrap();
+    let mut expected_batches = (&mut checkpoint_data).map(Result::unwrap).collect_vec();
+
+    // Ignore the v2 checkpointMetadata batch, because it's not actually part of replay
+    assert_eq!(expected_batches.pop().unwrap().data.len(), 1);
+
+    // Two-phase log replay
+    let mut phase1 = checkpoint_writer.phase1_checkpoint_data(&engine).unwrap();
+    let mut two_phase_batches = (&mut phase1).map(Result::unwrap).collect_vec();
```

Comment:
I might be misunderstand the code, but IIUC this requires `collect_vec()` to be called exactly once in order for the call below to `finish()` to work correctly?  If so is this a common idiom in Rust/Kernel?